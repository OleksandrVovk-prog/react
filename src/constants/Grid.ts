export const GRID_WIDTH = 50;
export const GRID_HEIGHT = 50;
