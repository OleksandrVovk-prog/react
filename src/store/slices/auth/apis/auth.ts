import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';

import AUTHENTICATED_USER from '../tags';

import IAuthResponse from '../interfaces/IAuthResponse';
import IAuthRequest from '../interfaces/IAuthRequest';

const authApi = createApi({
  reducerPath: 'authApi',
  baseQuery: fetchBaseQuery({
    baseUrl: process.env.REACT_APP_API,
  }),
  tagTypes: [AUTHENTICATED_USER],
  endpoints: (builder) => ({
    login: builder.mutation<IAuthResponse, Partial<IAuthRequest>>({
      invalidatesTags: [AUTHENTICATED_USER],
      query: (user) => ({
        url: '/auth/login',
        method: 'POST',
        body: user,
      }),
    }),
    fetchUser: builder.query<IAuthResponse, number | undefined>({
      providesTags: [AUTHENTICATED_USER],
      query: (id) => ({
        url: `/users/${id}`,
        method: 'GET',
      }),
    }),
  }),
});

export const { useLoginMutation, useFetchUserQuery } = authApi;
export default authApi;
