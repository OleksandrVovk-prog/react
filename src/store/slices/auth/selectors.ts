import type { RootState } from '../../types/TStore';

import IAuthResponse from './interfaces/IAuthResponse';

export const selectUserId = (state: RootState): IAuthResponse['id'] => state.auth.id;

export const selectUserToken = (state: RootState): IAuthResponse['token'] => state.auth.token;
