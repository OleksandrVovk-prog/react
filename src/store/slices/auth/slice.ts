import { createSlice } from '@reduxjs/toolkit';
// import dummyApi from '../../apis/dummy';
// import authApi from './apis/auth';
import dummyAuthApi from './apis/dummyAuth';

import IAuthResponse from './interfaces/IAuthResponse';

const initialState: IAuthResponse = {
  id: undefined,
  token: undefined,
};

const userSlice = createSlice({
  name: 'user',
  initialState,
  reducers: {
    logout: () => initialState,
  },
  extraReducers: (builder): void => {
    builder.addMatcher(
      dummyAuthApi.endpoints.login.matchFulfilled,
      (state, { payload }): void => {
        state.id = payload.id;
        state.token = payload.token;
      },
    );
  },
});

export const { logout } = userSlice.actions;

export default userSlice.reducer;
