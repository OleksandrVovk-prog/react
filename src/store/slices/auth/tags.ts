const AUTHENTICATED_USER = 'AUTHENTICATED_USER';

export default AUTHENTICATED_USER;
