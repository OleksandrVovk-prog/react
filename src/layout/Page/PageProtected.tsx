import { Navigate } from 'react-router-dom';
import { useAppSelector } from '../../store/hooks/useApp';
import { selectUserId } from '../../store/slices/auth/selectors';
import Page from './Page';

import IPageProtected from './interfaces/IPageProtected';

/**
 * Protected page layout
 */
function PageProtected({ redirectUri }: IPageProtected): JSX.Element {
  const id = useAppSelector(selectUserId);
  return id ? (
    <Page />
  ) : (
    <Navigate to={redirectUri} />
  );
}

export default PageProtected;
