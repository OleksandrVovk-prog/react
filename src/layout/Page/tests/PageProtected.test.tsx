import { screen } from '@testing-library/react';
import render from '../../../../__tests__/test-utils';
import { mockedNavigateId } from '../../../../__tests__/constants';

import PageProtected from '../PageProtected';
import auth from '../../../mocks/auth';

describe('<PageProtected />', () => {
  test('should render component', async () => {
    const { asFragment } = render(<PageProtected redirectUri="/login" />, { preloadedState: { auth } });
    expect(asFragment()).toMatchSnapshot();
  });
  test('should stay on page', async () => {
    render(<PageProtected redirectUri="/login" />, { preloadedState: { auth } });
    expect(screen.queryByTestId(mockedNavigateId)).not.toBeInTheDocument();
  });
  test('should redirect to Login', async () => {
    render(<PageProtected redirectUri="/login" />);
    expect(screen.getByTestId(mockedNavigateId)).toBeInTheDocument();
  });
});
