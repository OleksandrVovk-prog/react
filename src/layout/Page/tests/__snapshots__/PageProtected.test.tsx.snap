// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PageProtected /> should render component 1`] = `
<DocumentFragment>
  <div
    class="page"
  >
    <header
      class="header"
      data-testid="header"
    >
      <h2
        class="appSubtitle"
        data-testid="subtitle"
      >
        Try to build something great
      </h2>
      <nav
        class="headerNav"
      >
        <a
          class="headerNavLink"
          href="/"
        >
          Home
        </a>
        <a
          class="headerNavLink"
          href="/about"
        >
          About
        </a>
      </nav>
      <div
        class="toggleLanguage"
      >
        <div>
          <img
            alt="loader"
            class="appLoader"
            src="loader.svg"
          />
        </div>
        <button
          data-testid="toggle-language-button"
          type="button"
        >
          EN
        </button>
      </div>
    </header>
    <h1
      class="appTitle"
      data-testid="main-title"
    >
      Home
    </h1>
  </div>
</DocumentFragment>
`;
